src/web/web_base.nit

```diff
@@ -64,3 +68,148 @@ redef class HttpRequest
 	# Checks the URL get parameter `?json=true`.
 	fun is_json_asked: Bool do return bool_arg("json") or else false
 end
+
+redef class MEntity
+
+	# URL to `self` within the web interface.
+	fun web_url: String is abstract
+
+	# URL to `self` within the JSON api.
+	fun api_url: String do return "/api/entity/" / full_name
+
+	redef fun json do
+		var obj = super
+		obj["web_url"] = web_url
+		obj["api_url"] = api_url
+		return obj
+	end
+
+	# Get the full json repesentation of `self` with MEntityRefs resolved.
+	fun api_json(handler: ModelHandler): JsonObject do return json
+end
+
+redef class MEntityRef
+	redef fun json do
+		var obj = super
+		obj["web_url"] = mentity.web_url
+		obj["api_url"] = mentity.api_url
+		obj["name"] = mentity.name
+		obj["mdoc"] = mentity.mdoc_or_fallback
+		obj["visibility"] = mentity.visibility
+		obj["location"] = mentity.location
+		var modifiers = new JsonArray
+		for modifier in mentity.collect_modifiers do
+			modifiers.add modifier
+		end
+		obj["modifiers"] = modifiers
+		return obj
+	end
+end
+
+redef class MDoc
+
+	# Add doc down processing
+	redef fun json do
+		var obj = super
+		obj["synopsis"] = synopsis
+		obj["documentation"] = documentation
+		obj["comment"] = comment
+		obj["html_synopsis"] = html_synopsis.write_to_string
+		obj["html_documentation"] = html_documentation.write_to_string
+		obj["html_comment"] = html_comment.write_to_string
+		return obj
+	end
+end
+
+redef class MPackage
+	redef var web_url = "/package/{full_name}" is lazy
+end
+
+redef class MGroup
+	redef var web_url = "/group/{full_name}" is lazy
+end
+
+redef class MModule
+	redef var web_url = "/module/{full_name}" is lazy
```

Comment:
I admit it, maybe this feature comes too soon, this is for the views.

I use the kind in the view to provide different displays depending on the route.